// @elysiajs/node/dist/cjs/handle.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/handle.ts
var handle_exports = {};
__export(handle_exports, {
  createStaticHandler: () => createStaticHandler,
  errorToResponse: () => errorToResponse,
  mapCompactResponse: () => mapCompactResponse,
  mapEarlyResponse: () => mapEarlyResponse,
  mapResponse: () => mapResponse
});
module.exports = __toCommonJS(handle_exports);
var import_srvx2 = require("srvx");
var import_utils3 = require("elysia/adapter/utils");

// src/utils.ts
var import_srvx = require("srvx");
var import_utils = require("elysia/utils");
var import_utils2 = require("elysia/adapter/utils");
var handleFile = (response, set2) => {
  if (response instanceof Promise)
    return response.then((res) => handleFile(res, set2));
  const size = response.size;
  const immutable = set2 && (set2.status === 206 || set2.status === 304 || set2.status === 412 || set2.status === 416);
  const defaultHeader = immutable ? {
    "transfer-encoding": "chunked"
  } : {
    "accept-ranges": "bytes",
    "content-range": size ? `bytes 0-${size - 1}/${size}` : void 0,
    "transfer-encoding": "chunked"
  };
  if (!set2 && !size) return new import_srvx.FastResponse(response);
  if (!set2)
    return new import_srvx.FastResponse(response, {
      headers: defaultHeader
    });
  if (set2.headers instanceof Headers) {
    let setHeaders = defaultHeader;
    setHeaders = {};
    for (const [key, value] of set2.headers.entries())
      if (key in set2.headers) setHeaders[key] = value;
    if (immutable) {
      delete set2.headers["content-length"];
      delete set2.headers["accept-ranges"];
    }
    return new import_srvx.FastResponse(response, {
      status: set2.status,
      headers: setHeaders
    });
  }
  if ((0, import_utils.isNotEmpty)(set2.headers))
    return new import_srvx.FastResponse(response, {
      status: set2.status,
      headers: Object.assign(defaultHeader, set2.headers)
    });
  return new import_srvx.FastResponse(response, {
    status: set2.status,
    headers: defaultHeader
  });
};
function mergeHeaders(responseHeaders, setHeaders) {
  const headers = new Headers(Object.fromEntries(responseHeaders.entries()));
  if (setHeaders instanceof Headers)
    for (const key of setHeaders.keys()) {
      if (key === "set-cookie") {
        if (headers.has("set-cookie")) continue;
        for (const cookie of setHeaders.getSetCookie())
          headers.append("set-cookie", cookie);
      } else if (!responseHeaders.has(key))
        headers.set(key, setHeaders?.get(key) ?? "");
    }
  else
    for (const key in setHeaders)
      if (key === "set-cookie")
        headers.append(key, setHeaders[key]);
      else if (!responseHeaders.has(key))
        headers.set(key, setHeaders[key]);
  return headers;
}
function mergeStatus(responseStatus, setStatus) {
  if (typeof setStatus === "string") setStatus = import_utils.StatusMap[setStatus];
  if (responseStatus === 200) return setStatus;
  return responseStatus;
}
var createResponseHandler = (handler) => {
  const handleStream2 = (0, import_utils2.createStreamHandler)(handler);
  return (response, set2, request) => {
    const newResponse = new import_srvx.FastResponse(response.body, {
      headers: mergeHeaders(response.headers, set2.headers),
      status: mergeStatus(response.status, set2.status)
    });
    if (!newResponse.headers.has("content-length") && newResponse.headers.get("transfer-encoding") === "chunked")
      return handleStream2(
        (0, import_utils2.streamResponse)(newResponse),
        (0, import_utils2.responseToSetHeaders)(newResponse, set2),
        request,
        // @ts-ignore
        true
        // don't auto-format SSE for pre-formatted Response
      );
    return newResponse;
  };
};

// src/handle.ts
var import_file = require("elysia/universal/file");
var import_utils5 = require("elysia/utils");
var import_cookies = require("elysia/cookies");
var import_error = require("elysia/error");
var handleElysiaFile = (file, set2 = {
  headers: {}
}) => {
  const path = file.path;
  const contentType = import_file.mime[path.slice(path.lastIndexOf(".") + 1)];
  if (contentType) set2.headers["content-type"] = contentType;
  if (file.stats && set2 && set2.status !== 206 && set2.status !== 304 && set2.status !== 412 && set2.status !== 416)
    return file.stats.then((stat) => {
      const size = stat.size;
      if (size !== void 0) {
        set2.headers["content-range"] = `bytes 0-${size - 1}/${size}`;
        set2.headers["content-length"] = size;
      }
      return handleFile(file.value, set2);
    });
  return handleFile(file.value, set2);
};
var mapResponse = (response, set2, request) => {
  if ((0, import_utils5.isNotEmpty)(set2.headers) || set2.status !== 200 || set2.cookie) {
    (0, import_utils3.handleSet)(set2);
    switch (response?.constructor?.name) {
      case "String":
        set2.headers["content-type"] = "text/plain";
        return new import_srvx2.FastResponse(response, set2);
      case "Array":
      case "Object":
        set2.headers["content-type"] = "application/json";
        return new import_srvx2.FastResponse(JSON.stringify(response), set2);
      case "ElysiaFile":
        return handleElysiaFile(response, set2);
      case "File":
        return handleFile(response, set2);
      case "Blob":
        return handleFile(response, set2);
      case "ElysiaCustomStatusResponse":
        set2.status = response.code;
        return mapResponse(
          response.response,
          set2,
          request
        );
      case "ReadableStream":
        if (!set2.headers["content-type"]?.startsWith(
          "text/event-stream"
        ))
          set2.headers["content-type"] = "text/event-stream; charset=utf-8";
        request?.signal?.addEventListener(
          "abort",
          {
            handleEvent() {
              if (request?.signal && !request?.signal?.aborted)
                response.cancel();
            }
          },
          {
            once: true
          }
        );
        return new import_srvx2.FastResponse(response, set2);
      case void 0:
        if (!response) return new import_srvx2.FastResponse(null, set2);
        return new import_srvx2.FastResponse(JSON.stringify(response), set2);
      case "Response":
        return handleResponse(response, set2, request);
      case "Error":
        return errorToResponse(response, set2);
      case "Promise":
        return response.then(
          (x) => mapResponse(x, set2, request)
        );
      case "Function":
        return mapResponse(response(), set2, request);
      case "Number":
      case "Boolean":
        return new import_srvx2.FastResponse(
          response.toString(),
          set2
        );
      case "Cookie":
        if (response instanceof import_cookies.Cookie)
          return new import_srvx2.FastResponse(response.value, set2);
        return new import_srvx2.FastResponse(response?.toString(), set2);
      case "FormData":
        return new import_srvx2.FastResponse(response, set2);
      default:
        if (response instanceof import_srvx2.FastResponse)
          return handleResponse(response, set2, request);
        if (response instanceof Promise)
          return response.then((x) => mapResponse(x, set2));
        if (response instanceof Error)
          return errorToResponse(response, set2);
        if (response instanceof import_error.ElysiaCustomStatusResponse) {
          set2.status = response.code;
          return mapResponse(
            response.response,
            set2,
            request
          );
        }
        if (typeof response?.next === "function")
          return handleStream(response, set2, request);
        if (typeof response?.then === "function")
          return response.then((x) => mapResponse(x, set2));
        if (typeof response?.toResponse === "function")
          return mapResponse(response.toResponse(), set2);
        if ("charCodeAt" in response) {
          const code = response.charCodeAt(0);
          if (code === 123 || code === 91) {
            if (!set2.headers["Content-Type"])
              set2.headers["Content-Type"] = "application/json";
            return new import_srvx2.FastResponse(
              JSON.stringify(response),
              set2
            );
          }
        }
        return new import_srvx2.FastResponse(response, set2);
    }
  }
  if (response instanceof import_srvx2.FastResponse && !response.headers.has("content-length") && response.headers.get("transfer-encoding") === "chunked")
    return handleStream(
      (0, import_utils3.streamResponse)(response),
      (0, import_utils3.responseToSetHeaders)(response, set2),
      request
    );
  if (
    // @ts-expect-error
    typeof response?.next === "function" || response instanceof ReadableStream
  )
    return handleStream(response, set2, request);
  return mapCompactResponse(response, request);
};
var mapEarlyResponse = (response, set2, request) => {
  if (response === void 0 || response === null) return;
  if ((0, import_utils5.isNotEmpty)(set2.headers) || set2.status !== 200 || set2.cookie) {
    (0, import_utils3.handleSet)(set2);
    switch (response?.constructor?.name) {
      case "String":
        set2.headers["content-type"] = "text/plain";
        return new import_srvx2.FastResponse(response, set2);
      case "Array":
      case "Object":
        set2.headers["content-type"] = "application/json";
        return new import_srvx2.FastResponse(JSON.stringify(response), set2);
      case "ElysiaFile":
        return handleElysiaFile(response, set2);
      case "File":
        return handleFile(response, set2);
      case "Blob":
        return handleFile(response, set2);
      case "ElysiaCustomStatusResponse":
        set2.status = response.code;
        return mapEarlyResponse(
          response.response,
          set2,
          request
        );
      case "ReadableStream":
        if (!set2.headers["content-type"]?.startsWith(
          "text/event-stream"
        ))
          set2.headers["content-type"] = "text/event-stream; charset=utf-8";
        request?.signal?.addEventListener(
          "abort",
          {
            handleEvent() {
              if (request?.signal && !request?.signal?.aborted)
                response.cancel();
            }
          },
          {
            once: true
          }
        );
        return new import_srvx2.FastResponse(response, set2);
      case void 0:
        if (!response) return;
        return new import_srvx2.FastResponse(JSON.stringify(response), set2);
      case "Response":
        return handleResponse(response, set2, request);
      case "Promise":
        return response.then(
          (x) => mapEarlyResponse(x, set2)
        );
      case "Error":
        return errorToResponse(response, set2);
      case "Function":
        return mapEarlyResponse(response(), set2);
      case "Number":
      case "Boolean":
        return new import_srvx2.FastResponse(
          response.toString(),
          set2
        );
      case "FormData":
        return new import_srvx2.FastResponse(response);
      case "Cookie":
        if (response instanceof import_cookies.Cookie)
          return new import_srvx2.FastResponse(response.value, set2);
        return new import_srvx2.FastResponse(response?.toString(), set2);
      default:
        if (response instanceof import_srvx2.FastResponse)
          return handleResponse(response, set2, request);
        if (response instanceof Promise)
          return response.then((x) => mapEarlyResponse(x, set2));
        if (response instanceof Error)
          return errorToResponse(response, set2);
        if (response instanceof import_error.ElysiaCustomStatusResponse) {
          set2.status = response.code;
          return mapEarlyResponse(
            response.response,
            set2,
            request
          );
        }
        if (typeof response?.next === "function")
          return handleStream(response, set2, request);
        if (typeof response?.then === "function")
          return response.then((x) => mapEarlyResponse(x, set2));
        if (typeof response?.toResponse === "function")
          return mapEarlyResponse(response.toResponse(), set2);
        if ("charCodeAt" in response) {
          const code = response.charCodeAt(0);
          if (code === 123 || code === 91) {
            if (!set2.headers["Content-Type"])
              set2.headers["Content-Type"] = "application/json";
            return new import_srvx2.FastResponse(
              JSON.stringify(response),
              set2
            );
          }
        }
        return new import_srvx2.FastResponse(response, set2);
    }
  } else
    switch (response?.constructor?.name) {
      case "String":
        set2.headers["content-type"] = "text/plain";
        return new import_srvx2.FastResponse(response);
      case "Array":
      case "Object":
        set2.headers["content-type"] = "application/json";
        return new import_srvx2.FastResponse(JSON.stringify(response), set2);
      case "ElysiaFile":
        return handleElysiaFile(response, set2);
      case "File":
        return handleFile(response, set2);
      case "Blob":
        return handleFile(response, set2);
      case "ElysiaCustomStatusResponse":
        set2.status = response.code;
        return mapEarlyResponse(
          response.response,
          set2,
          request
        );
      case "ReadableStream":
        request?.signal?.addEventListener(
          "abort",
          {
            handleEvent() {
              if (request?.signal && !request?.signal?.aborted)
                response.cancel();
            }
          },
          {
            once: true
          }
        );
        return new import_srvx2.FastResponse(response, {
          headers: {
            "Content-Type": "text/event-stream; charset=utf-8"
          }
        });
      case void 0:
        if (!response) return new import_srvx2.FastResponse("");
        return new import_srvx2.FastResponse(JSON.stringify(response), {
          headers: {
            "content-type": "application/json"
          }
        });
      case "Response":
        if (!response.headers.has("content-length") && response.headers.get("transfer-encoding") === "chunked")
          return handleStream(
            (0, import_utils3.streamResponse)(response),
            (0, import_utils3.responseToSetHeaders)(response),
            request
          );
        return response;
      case "Promise":
        return response.then((x) => {
          const r = mapEarlyResponse(x, set2);
          if (r !== void 0) return r;
        });
      case "Error":
        return errorToResponse(response, set2);
      case "Function":
        return mapCompactResponse(response(), request);
      case "Number":
      case "Boolean":
        return new import_srvx2.FastResponse(response.toString());
      case "Cookie":
        if (response instanceof import_cookies.Cookie)
          return new import_srvx2.FastResponse(response.value, set2);
        return new import_srvx2.FastResponse(response?.toString(), set2);
      case "FormData":
        return new import_srvx2.FastResponse(response);
      default:
        if (response instanceof import_srvx2.FastResponse) return response;
        if (response instanceof Promise)
          return response.then((x) => mapEarlyResponse(x, set2));
        if (response instanceof Error)
          return errorToResponse(response, set2);
        if (response instanceof import_error.ElysiaCustomStatusResponse) {
          set2.status = response.code;
          return mapEarlyResponse(
            response.response,
            set2,
            request
          );
        }
        if (typeof response?.next === "function")
          return handleStream(response, set2, request);
        if (typeof response?.then === "function")
          return response.then((x) => mapEarlyResponse(x, set2));
        if (typeof response?.toResponse === "function")
          return mapEarlyResponse(response.toResponse(), set2);
        if ("charCodeAt" in response) {
          const code = response.charCodeAt(0);
          if (code === 123 || code === 91) {
            if (!set2.headers["Content-Type"])
              set2.headers["Content-Type"] = "application/json";
            return new import_srvx2.FastResponse(
              JSON.stringify(response),
              set2
            );
          }
        }
        return new import_srvx2.FastResponse(response);
    }
};
var mapCompactResponse = (response, request) => {
  switch (response?.constructor?.name) {
    case "String":
      return new import_srvx2.FastResponse(response, {
        headers: {
          "Content-Type": "text/plain"
        }
      });
    case "Object":
    case "Array":
      return new import_srvx2.FastResponse(JSON.stringify(response), {
        headers: {
          "Content-Type": "application/json"
        }
      });
    case "ElysiaFile":
      return handleElysiaFile(response);
    case "File":
      return handleFile(response);
    case "Blob":
      return handleFile(response);
    case "ElysiaCustomStatusResponse":
      return mapResponse(
        response.response,
        {
          status: response.code,
          headers: {}
        }
      );
    case "ReadableStream":
      request?.signal?.addEventListener(
        "abort",
        {
          handleEvent() {
            if (request?.signal && !request?.signal?.aborted)
              response.cancel();
          }
        },
        {
          once: true
        }
      );
      return new import_srvx2.FastResponse(response, {
        headers: {
          "Content-Type": "text/event-stream; charset=utf-8"
        }
      });
    case void 0:
      if (!response) return new import_srvx2.FastResponse("");
      return new import_srvx2.FastResponse(JSON.stringify(response), {
        headers: {
          "content-type": "application/json"
        }
      });
    case "Response":
      if (response.headers.get("transfer-encoding") === "chunked")
        return handleStream(
          (0, import_utils3.streamResponse)(response),
          (0, import_utils3.responseToSetHeaders)(response),
          request
        );
      return response;
    case "Error":
      return errorToResponse(response);
    case "Promise":
      return response.then(
        (x) => mapCompactResponse(x, request)
      );
    // ? Maybe response or Blob
    case "Function":
      return mapCompactResponse(response(), request);
    case "Number":
    case "Boolean":
      return new import_srvx2.FastResponse(response.toString());
    case "FormData":
      return new import_srvx2.FastResponse(response);
    default:
      if (response instanceof import_srvx2.FastResponse) return response;
      if (response instanceof Promise)
        return response.then(
          (x) => mapCompactResponse(x, request)
        );
      if (response instanceof Error)
        return errorToResponse(response);
      if (response instanceof import_error.ElysiaCustomStatusResponse)
        return mapResponse(
          response.response,
          {
            status: response.code,
            headers: {}
          }
        );
      if (typeof response?.next === "function")
        return handleStream(response, void 0, request);
      if (typeof response?.then === "function")
        return response.then((x) => mapResponse(x, set));
      if (typeof response?.toResponse === "function")
        return mapCompactResponse(response.toResponse());
      if ("charCodeAt" in response) {
        const code = response.charCodeAt(0);
        if (code === 123 || code === 91) {
          return new import_srvx2.FastResponse(JSON.stringify(response), {
            headers: {
              "Content-Type": "application/json"
            }
          });
        }
      }
      return new import_srvx2.FastResponse(response);
  }
};
var errorToResponse = (error, set2) => new import_srvx2.FastResponse(
  JSON.stringify({
    name: error?.name,
    message: error?.message,
    cause: error?.cause
  }),
  {
    status: set2?.status !== 200 ? set2?.status ?? 500 : 500,
    headers: set2?.headers
  }
);
var createStaticHandler = (handle, hooks, setHeaders = {}) => {
  if (typeof handle === "function") return;
  const response = mapResponse(handle, {
    headers: setHeaders
  });
  if (!hooks.parse?.length && !hooks.transform?.length && !hooks.beforeHandle?.length && !hooks.afterHandle?.length)
    return response.clone.bind(response);
};
var handleResponse = createResponseHandler({
  mapResponse,
  mapCompactResponse
});
var handleStream = (0, import_utils3.createStreamHandler)({
  mapResponse,
  mapCompactResponse
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createStaticHandler,
  errorToResponse,
  mapCompactResponse,
  mapEarlyResponse,
  mapResponse
});
